packages/angular/cli/src/utilities/completion.ts

```diff
@@ -246,3 +270,38 @@ function getShellRunCommandCandidates(shell: string, home: string): string[] | u
     return undefined;
   }
 }
+
+/**
+ * Returns whether the user has a global CLI install or `undefined` if this can't be determined.
+ * Execution from `npx` is *not* considered a global CLI install.
+ *
+ * This does *not* mean the current execution is from a global CLI install, only that a global
+ * install exists on the system.
+ */
+export async function hasGlobalCliInstall(): Promise<boolean | undefined> {
+  // List all binaries with the `ng` name on the user's `$PATH`.
+  const proc = execFile('which', ['-a', 'ng']);
```

Comment:
Wouldn't it simpler to just use `exec`, this also ensures cross-platform compatibility.


```ts
function hasGlobalAngularCLI(): Promise<boolean> {
  return exec(`ng help`, {
    stdio: "ignore",
  })
    .then(() => true)
    .catch(() => false);
}

```

Reply:
I don't think that would be sufficient because we need to ignore any local installs of `ng`. Since NPM scripts add the local `node_modules/.bin/` to the `$PATH`, your example would return `true` without ever looking beyond the local scope.

`which -a` is necessary to list *all* the CLI binaries on the `$PATH`, whereby we can exclude local ones and determine if any global installs exist.

Reply:
Also updated this to properly parse `which` output as I realized the test was inadvertently relying on `which` failing due to the local CLI not being on the path, should be fixed now.

Reply:
> NPM scripts add the local `node_modules/.bin/` to the $PATH

AFAIK, this is old NPM behaviour prior to NPM 6 prior to NPX, and local `node_modules/.bin/` is no longer added to `$PATH`. 

I did try the above snippet, and locally I did work as expected, when `ng` is only installed locally it returned `false`, while when it is installed globally it return `true`.

```
$ which ng
ng not found

$ ls node_modules/.bin/ng
node_modules/.bin/ng

$ node --eval "require('child_process').execSync('ng help', { stdio: 'ignore' })"
child_process.js:866
    throw err;
    ^

Error: Command failed: ng help
    at checkExecSyncError (child_process.js:790:11)
    at Object.execSync (child_process.js:863:15)
    at [eval]:1:26
    at Script.runInThisContext (vm.js:134:12)
    at Object.runInThisContext (vm.js:310:38)
    at internal/process/execution.js:81:19
    at [eval]-wrapper:6:22
    at evalScript (internal/process/execution.js:80:60)
    at internal/main/eval_string.js:27:3 {
  status: 127,
  signal: null,
  output: [ null, null, null ],
  pid: 96183,
  stdout: null,
  stderr: null
}

$ npm i @angular/cli@next -g

added 206 packages, and audited 207 packages in 6s

25 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$  node --eval "require('child_process').execSync('ng help', { stdio: 'ignore' })"
```

I am not too strongly, about using `which`, but the code is more involved compared to executing `ng` directly.

Reply:
I don't think your test case of directly running `node --eval "require('child_process').execSync('ng help', { stdio: 'inherit' })"` in a project is accurate, Node doesn't modify the `$PATH` in this situation. But NPM does for NPM scripts, so if you do the same thing in script and execute that, a local CLI will successfully execute.

```jsonc
{
  "scripts": {
    "path-test": "node --eval \"require('child_process').execSync('ng help', { stdio: 'inherit' })\""
  },
  // ...
}
```

```
douglasparker@basilisk (~/Source/ng-new)
[22-05-18 10:13:48]$ node --eval "require('child_process').execSync('ng help', { stdio: 'inherit' })"
/bin/sh: line 1: ng: command not found

# Snip...
douglasparker@basilisk (~/Source/ng-new)
[22-05-18 10:13:51]$ npm run path-test

> ng-new@0.0.0 path-test
> node --eval "require('child_process').execSync('ng help', { stdio: 'inherit' })"

ng <command>

Commands:
  ng add <collection>            Adds support for an external library to your project.
  ng analytics                   Configures the gathering of Angular CLI usage metrics. See https://angular.io/cli/usage-analytics-gathering
  ng build [project]             Compiles an Angular application or library into an output directory named dist/ at the given output path.                                           [aliases: b]

# Snip...
```

Since the CLI can be run in this script context (such as `npm run build` or `npm start`), we have to assume any CLI on the `$PATH` could originate from a local install, meaning we need `which -a` to filter them out.

Reply:
I'll merge this now just so we can get in for the RC today, but I'm open to simplifying this if we can find a way that doesn't rely on `which -a`.